### PR TITLE
Clean up logic with task names to use persistent state

### PR DESCRIPTION
**Pull request by elingg**
This will fix the bug in Task Reconciliation.

### CODE REVIEW

**Comment by elingg** [comment]
@adam-mesos, please take a look at your earliest convenience. Thanks!

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 1]
```diff
@@ -452,7 +454,7 @@ private boolean tryToLaunchNameNode(SchedulerDriver driver, Offer offer) {
     List<String> deadNameNodes = persistentState.getDeadNameNodes();
 
     if (deadNameNodes.isEmpty()) {
-      if (liveState.getNameNodeSize() == HDFSConstants.TOTAL_NAME_NODES) {
+      if (persistentState.getNameNodes().size() == HDFSConstants.TOTAL_NAME_NODES) {
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: this contacts the zkState every time. What is the purpose of using persistent state over live state?

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 2]
```diff
@@ -416,7 +418,7 @@ private boolean tryToLaunchJournalNode(SchedulerDriver driver, Offer offer) {
     log.info(deadJournalNodes);
 
     if (deadJournalNodes.isEmpty()) {
-      if (liveState.getJournalNodeSize() == conf.getJournalNodeCount()) {
+      if (persistentState.getJournalNodes().size() == conf.getJournalNodeCount()) {
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: this contacts the zkState every time. What is the purpose of using persistent state over live state?

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java [hunk 3]
```diff
@@ -173,17 +183,23 @@ private void removeDeadDataNodes() {
...
 
-  public void addHdfsNode(Protos.TaskID taskId, String hostname, String taskType) {
+  public void addHdfsNode(Protos.TaskID taskId, String hostname, String taskType, String taskName) {
     switch (taskType) {
       case HDFSConstants.NAME_NODE_ID :
         HashMap<String, String> nameNodes = getNameNodes();
         nameNodes.put(hostname, taskId.getValue());
         setNameNodes(nameNodes);
+        HashMap<String, String> nameNodeTaskNames = getNameNodeTaskNames();
+        nameNodeTaskNames.put(taskId.getValue(), taskName);
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: lets use the task name as the key instead. That way we don't have to do `.values()` and we reinforce the idea of unique task name

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java [hunk 4]
```diff
@@ -173,17 +183,23 @@ private void removeDeadDataNodes() {
...
+        setNameNodeTaskNames(nameNodeTaskNames);
         break;
       case HDFSConstants.JOURNAL_NODE_ID :
         HashMap<String, String> journalNodes = getJournalNodes();
         journalNodes.put(hostname, taskId.getValue());
         setJournalNodes(journalNodes);
+        HashMap<String, String> journalNodeTaskNames = getJournalNodeTaskNames();
+        journalNodeTaskNames.put(taskId.getValue(), taskName);
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: lets use the task name as the key instead. That way we don't have to do `.values()` and we reinforce the idea of unique task name

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 5]
```diff
@@ -284,17 +283,20 @@ private boolean launchNode(SchedulerDriver driver, Offer offer,
...
+      liveState.addStagingTask(task.getTaskId());
+      persistentState.addHdfsNode(taskId, offer.getHostname(), taskType, taskName);
     }
     driver.launchTasks(Arrays.asList(offer.getId()), tasks);
     return true;
   }
 
   private String getNextTaskType(String taskType) {
+    Collection<String> nameNodeTaskNames = persistentState.getNameNodeTaskNames().values();
+    Collection<String> journalNodeTaskNames = persistentState.getJournalNodeTaskNames().values();
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: lets switch the keys and values and then do a `.keySet()`, which is more efficient.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 1 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: We need to persist the data in Zookeeper.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 2 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: We need to persist the data in Zookeeper to make sure reconciliation works.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 1 above)
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: but for efficiency, the livestate is effectively the same since we reconcile tasks.

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java (hunk 3 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: We need to remove by taskId because that's what we know when the task is removed so this way is more convenient. Even if we reversed this map, we would have to call keys instead of values.

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java (hunk 4 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: see above

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 5 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: See above

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 1 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: Hm, we actually moved this to be stored in PersistentState instead of LiveState. This makes sense because we do need to persist this for reconciliation. We already store the taskid and hostname in the persistent state. We are simply adding the taskname here.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 6]
```diff
@@ -284,17 +283,20 @@ private boolean launchNode(SchedulerDriver driver, Offer offer,
...
-      liveState.addStagingTask(task.getTaskId(), taskName);
-      persistentState.addHdfsNode(taskId, offer.getHostname(), taskType);
+      liveState.addStagingTask(task.getTaskId());
+      persistentState.addHdfsNode(taskId, offer.getHostname(), taskType, taskName);
     }
     driver.launchTasks(Arrays.asList(offer.getId()), tasks);
     return true;
   }
 
   private String getNextTaskType(String taskType) {
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: this should be getNextTaskName

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 7]
```diff
@@ -284,17 +283,20 @@ private boolean launchNode(SchedulerDriver driver, Offer offer,
...
+    Collection<String> nameNodeTaskNames = persistentState.getNameNodeTaskNames().values();
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: It would be better to put these inside the if statement so we're not contacting the zkState for both every time.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 8]
```diff
@@ -284,17 +283,20 @@ private boolean launchNode(SchedulerDriver driver, Offer offer,
...
+    Collection<String> journalNodeTaskNames = persistentState.getJournalNodeTaskNames().values();
+
     if (taskType.equals(HDFSConstants.NAME_NODE_ID)) {
       for (int i = 1; i <= HDFSConstants.TOTAL_NAME_NODES; i++) {
-        if (!liveState.getNameNodeNames().containsValue(HDFSConstants.NAME_NODE_ID + i)) {
+        if (!nameNodeTaskNames.contains(HDFSConstants.NAME_NODE_ID + i)) {
           return HDFSConstants.NAME_NODE_ID + i;
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: Since we're never supposed to reach here when we've launched enough tasks, lets return an error instead of ""

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 9]
```diff
@@ -302,7 +304,7 @@ private String getNextTaskType(String taskType) {
     }
     if (taskType.equals(HDFSConstants.JOURNAL_NODE_ID)) {
       for (int i = 1; i <= conf.getJournalNodeCount(); i++) {
-        if (!liveState.getJournalNodeNames().containsValue(HDFSConstants.JOURNAL_NODE_ID + i)) {
+        if (!journalNodeTaskNames.contains(HDFSConstants.JOURNAL_NODE_ID + i)) {
           return HDFSConstants.JOURNAL_NODE_ID + i;
```
Comment (abhay-agarwal) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: Since we're never supposed to reach here when we've launched enough tasks, lets return an error instead of ""

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 7 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: Great point. Will fix this.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 6 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: yes taskName is "String taskName = getNextTaskType(taskType);"

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 8 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: Yes, good point will do.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 9 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: See above.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 6 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: oh, got it, it should be renamed. will rename it.

**Comment by nicgrayson** [comment]
@elingg what bug in reconciliation? Is there an issue for it? 

Also have you considered using curator to have the zk calls be less taxing?

**Comment by abhay-agarwal** [comment]
@nicgrayson its #115

**Comment by elingg** [comment]
@nicgrayson, @abhay-agarwal sent you the issue. It seems some aspects of reconciliation got broken with the addition to task names, but this JIRA should fix it. Re curator, yes that could be one possibility to avoid calls to ZK. We might also have some cache. We currently have this issue for all of the persistent state.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java [hunk 10]
```diff
@@ -268,8 +268,7 @@ private boolean launchNode(SchedulerDriver driver, Offer offer,
     List<TaskInfo> tasks = new ArrayList<>();
     for (String taskType : taskTypes) {
       List<Resource> taskResources = getTaskResources(taskType);
-      String taskName = getNextTaskType(taskType);
-      if (taskName.isEmpty()) return false;
+      String taskName = getNextTaskName(taskType);
```
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: Since getNextTaskName can now throw an exception (instead of returning empty), should we try/catch that exception here, or is it fine to let it propagate up?

Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: It's a runtime exception so I think should be safe to propagate up. It follows the same pattern as the other exceptions in scheduler.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 2 above)
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: For these getters we could cache the state locally, so that we only have to pull from zk on scheduler failover/recovery. We'll still push updates out to ZK asap so that the state is up to date in the event of failover. This is probably worth a separate issue/PR.

File: src/main/java/org/apache/mesos/hdfs/Scheduler.java (hunk 1 above)
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/hdfs/Scheduler.java]: I think LiveState should just be a (write-through) cache of the PersistentState. That way the scheduler only interacts with one state abstraction, which then decides when to write to zk (on any state change) or when to read from zk (on startup/recovery).

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java (hunk 3 above)
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: Should we verify that there is no preexisting entry for that taskId? Otherwise, we would just overwrite the value. Probably won't happen, but worth an assertion check.

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java [hunk 11]
```diff
@@ -201,11 +217,14 @@ public void addHdfsNode(Protos.TaskID taskId, String hostname, String taskType)
   // Possibly call removeTask(slaveId, taskId) to avoid iterating through all maps
   public void removeTaskId(String taskId) {
     HashMap<String, String> journalNodes = getJournalNodes();
+    HashMap<String, String> journalNodeTaskNames = getJournalNodeTaskNames();
```
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: At least as long as this pulls directly from zk, let's not call this until we're inside the if {} block where we know we'll actually use it, right before the remove call.

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java [hunk 12]
```diff
@@ -266,6 +288,22 @@ private void setJournalNodes(HashMap<String, String> journalNodes) {
     }
   }
 
+  private void setNameNodeTaskNames(HashMap<String, String> nameNodeTaskNames) {
+    try {
+      set(NAMENODE_TASKNAMES_KEY, nameNodeTaskNames);
+    } catch (Exception e) {
+      log.error("Error while setting namenodes in persistent state", e);
```
Comment (adam-mesos) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: Is logging an error the best/only reaction to an exception? Maybe retry, or propagate up?

**Comment by adam-mesos** [comment]
LGTM. A couple of minor points about error-handling/propagation and moving a zk get() further into the if branch where it's actually used. Address those, and then :rocket: 
I also think we should turn LiveState into a write-through cache for PersistentState, to restrict the scheduler to a single state abstraction. But that can be a separate issue/PR.

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java (hunk 11 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: Done

File: src/main/java/org/apache/mesos/hdfs/state/PersistentState.java (hunk 12 above)
Comment (elingg) [inline src/main/java/org/apache/mesos/hdfs/state/PersistentState.java]: You are correct. I think we need better error handling in persistent state. Propagating the exception up will be better, but that will be a refactor of the structure of PersistentState and probably should be another PR. Meanwhile, in most cases, if there is an error here, it will be caught elsewhere. The framework should be able to react.
